File: api/sos/sun.py
"""Sunrise and sunset from the NOAA solar calculator equations: pure arithmetic, no data files and no network.

A port of `web/src/tools/sun.ts` so the engine can decide whether it is dark at home without asking the browser.
Times are timezone-aware UTC datetimes; the caller formats them in the box's zone."""
from __future__ import annotations

import math
from datetime import date, datetime, timedelta, timezone

RAD = math.pi / 180.0
ZENITH_SUN = 90.833          # the sun's centre at rise and set, allowing for refraction and its radius
ZENITH_CIVIL = 96.0


def _julian_day(moment: datetime) -> float:
    return moment.timestamp() / 86400.0 + 2440587.5


def _solar_params(jd: float) -> tuple[float, float]:
    """Equation of time (minutes) and the solar declination (degrees) for a Julian day."""
    t = (jd - 2451545.0) / 36525.0
    l0 = (280.46646 + t * (36000.76983 + t * 0.0003032)) % 360.0
    m = 357.52911 + t * (35999.05029 - 0.0001537 * t)
    e = 0.016708634 - t * (0.000042037 + 0.0000001267 * t)
    c = (math.sin(m * RAD) * (1.914602 - t * (0.004817 + 0.000014 * t))
         + math.sin(2 * m * RAD) * (0.019993 - 0.000101 * t)
         + math.sin(3 * m * RAD) * 0.000289)
    true_long = l0 + c
    omega = 125.04 - 1934.136 * t
    apparent_long = true_long - 0.00569 - 0.00478 * math.sin(omega * RAD)
    obliq0 = 23 + (26 + (21.448 - t * (46.815 + t * (0.00059 - t * 0.001813))) / 60) / 60
    obliq = obliq0 + 0.00256 * math.cos(omega * RAD)
    declination = math.asin(math.sin(obliq * RAD) * math.sin(apparent_long * RAD)) / RAD
    y = math.tan((obliq / 2) * RAD) ** 2
    eq_time = 4 / RAD * (
        y * math.sin(2 * l0 * RAD) - 2 * e * math.sin(m * RAD) + 4 * e * y * math.sin(m * RAD) * math.cos(2 * l0 * RAD)
        - 0.5 * y * y * math.sin(4 * l0 * RAD) - 1.25 * e * e * math.sin(2 * m * RAD))
    return eq_time, declination


def _hour_angle(lat: float, declination: float, zenith: float) -> float | None:
    """Degrees from solar noon at which the sun reaches the zenith, or None on a polar day or night."""
    cos_h = (math.cos(zenith * RAD) / (math.cos(lat * RAD) * math.cos(declination * RAD))
             - math.tan(lat * RAD) * math.tan(declination * RAD))
    if cos_h > 1 or cos_h < -1:
        return None
    return math.acos(cos_h) / RAD


def _at(day_start: datetime, minutes: float) -> datetime:
    return day_start + timedelta(minutes=minutes)


def sun_times(lat: float, lon: float, day: date | datetime) -> tuple[datetime | None, datetime | None]:
    """Sunrise and sunset in UTC for the UTC calendar day of `day`.

    Returns (None, None) on a polar day or night, when the sun neither rises nor sets."""
    if isinstance(day, datetime):
        day = day.astimezone(timezone.utc).date() if day.tzinfo else day.date()
    day_start = datetime(day.year, day.month, day.day, tzinfo=timezone.utc)
    noon_guess = _julian_day(day_start) + 0.5 - lon / 360.0
    eq0, _ = _solar_params(noon_guess)
    noon_min = 720 - 4 * lon - eq0
    solar_noon = _at(day_start, noon_min)
    eq_noon, dec_noon = _solar_params(_julian_day(solar_noon))
    ha = _hour_angle(lat, dec_noon, ZENITH_SUN)
    if ha is None:
        return None, None
    eq_rise, dec_rise = _solar_params(_julian_day(_at(day_start, noon_min - ha * 4)))
    eq_set, dec_set = _solar_params(_julian_day(_at(day_start, noon_min + ha * 4)))
    ha_rise = _hour_angle(lat, dec_rise, ZENITH_SUN) or ha
    ha_set = _hour_angle(lat, dec_set, ZENITH_SUN) or ha
    sunrise = _at(day_start, 720 - 4 * (lon + ha_rise) - eq_rise)
    sunset = _at(day_start, 720 - 4 * (lon - ha_set) - eq_set)
    return sunrise, sunset


def is_dark(lat: float, lon: float, now: datetime) -> bool:
    """True between sunset and sunrise. On a polar day it is never dark; on a polar night, always."""
    now = now if now.tzinfo else now.replace(tzinfo=timezone.utc)
    sunrise, sunset = sun_times(lat, lon, now)
    if sunrise is None or sunset is None:
        _, dec = _solar_params(_julian_day(now))
        return lat * dec < 0                       # the sun is over the other hemisphere: polar night
    return now < sunrise or now >= sunset

File: web/src/tools/sun.ts
// Sunrise, sunset and civil twilight from the NOAA solar calculator equations. Pure arithmetic, no data files.
// Times come back as UTC Date objects; callers format them in the viewer's zone.

export type SunTimes =
  | { polar: null; sunrise: Date; sunset: Date; civilDawn: Date | null; civilDusk: Date | null; dayLengthMin: number; solarNoon: Date }
  | { polar: 'day' | 'night'; solarNoon: Date; dayLengthMin: number };

const RAD = Math.PI / 180;

function julianDay(date: Date): number {
  return date.getTime() / 86_400_000 + 2_440_587.5;
}

/** Equation of time (minutes) and solar declination (degrees) for a Julian day. */
function solarParams(jd: number): { eqTime: number; declination: number } {
  const t = (jd - 2_451_545) / 36_525;
  const l0 = (280.46646 + t * (36_000.76983 + t * 0.0003032)) % 360;
  const m = 357.52911 + t * (35_999.05029 - 0.0001537 * t);
  const e = 0.016708634 - t * (0.000042037 + 0.0000001267 * t);
  const c = Math.sin(m * RAD) * (1.914602 - t * (0.004817 + 0.000014 * t)) + Math.sin(2 * m * RAD) * (0.019993 - 0.000101 * t) + Math.sin(3 * m * RAD) * 0.000289;
  const trueLong = l0 + c;
  const omega = 125.04 - 1934.136 * t;
  const apparentLong = trueLong - 0.00569 - 0.00478 * Math.sin(omega * RAD);
  const obliq0 = 23 + (26 + (21.448 - t * (46.815 + t * (0.00059 - t * 0.001813))) / 60) / 60;
  const obliq = obliq0 + 0.00256 * Math.cos(omega * RAD);
  const declination = Math.asin(Math.sin(obliq * RAD) * Math.sin(apparentLong * RAD)) / RAD;
  const y = Math.tan((obliq / 2) * RAD) ** 2;
  const eqTime = 4 / RAD * (
    y * Math.sin(2 * l0 * RAD) - 2 * e * Math.sin(m * RAD) + 4 * e * y * Math.sin(m * RAD) * Math.cos(2 * l0 * RAD)
    - 0.5 * y * y * Math.sin(4 * l0 * RAD) - 1.25 * e * e * Math.sin(2 * m * RAD));
  return { eqTime, declination };
}

/** Hour angle (degrees) at which the sun's centre reaches the given zenith, or null when it never does that day. */
function hourAngle(lat: number, declination: number, zenith: number): number | null {
  const cosH = (Math.cos(zenith * RAD) / (Math.cos(lat * RAD) * Math.cos(declination * RAD))) - Math.tan(lat * RAD) * Math.tan(declination * RAD);
  if (cosH > 1 || cosH < -1) return null;
  return Math.acos(cosH) / RAD;
}

function utcDate(dayStartUtc: Date, minutes: number): Date {
  return new Date(dayStartUtc.getTime() + minutes * 60_000);
}

/** Sun events for the civil date `date` (its UTC calendar day is used) at a location. */
export function sunTimes(lat: number, lon: number, date: Date): SunTimes {
  const dayStart = new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate()));
  // iterate once from noon so the declination matches the actual event times
  const noonGuess = julianDay(dayStart) + 0.5 - lon / 360;
  const p0 = solarParams(noonGuess);
  const noonMin = 720 - 4 * lon - p0.eqTime;
  const solarNoon = utcDate(dayStart, noonMin);
  const p = solarParams(julianDay(solarNoon));
  const ha = hourAngle(lat, p.declination, 90.833);
  if (ha === null) {
    const up = lat * p.declination > 0;   // same hemisphere as the sun: it never sets
    return { polar: up ? 'day' : 'night', solarNoon, dayLengthMin: up ? 1440 : 0 };
  }
  const rise = solarParams(julianDay(utcDate(dayStart, noonMin - ha * 4)));
  const set = solarParams(julianDay(utcDate(dayStart, noonMin + ha * 4)));
  const haRise = hourAngle(lat, rise.declination, 90.833) ?? ha;
  const haSet = hourAngle(lat, set.declination, 90.833) ?? ha;
  const sunriseMin = 720 - 4 * (lon + haRise) - rise.eqTime;
  const sunsetMin = 720 - 4 * (lon - haSet) - set.eqTime;
  const civil = hourAngle(lat, p.declination, 96);
  return {
    polar: null,
    sunrise: utcDate(dayStart, sunriseMin),
    sunset: utcDate(dayStart, sunsetMin),
    civilDawn: civil === null ? null : utcDate(dayStart, 720 - 4 * (lon + civil) - p.eqTime),
    civilDusk: civil === null ? null : utcDate(dayStart, 720 - 4 * (lon - civil) - p.eqTime),
    dayLengthMin: Math.round(sunsetMin - sunriseMin),
    solarNoon,
  };
}

export function formatDayLength(minutes: number): string {
  const h = Math.floor(minutes / 60);
  const m = Math.round(minutes % 60);
  return `${h} h ${m.toString().padStart(2, '0')} min`;
}
